### PR TITLE
internal user-interaction: Enable 2 consecutive test suites on seperate VOs #25

### PR DESCRIPTION
**Pull request by patrick-austin**
Enable 2 consecutive test suites on seperate VOs
------------------

Updated the multi-VO Travis suite to run twice on two VOs (`tst` and `ts2`) without clearing the DB (but the cache **is** cleared as otherwise we wouldn't pick up newly added RSEs - this is not an exclusively multi-VO issue),  along with some other modifications needed to make this work:

- Removed hardcoding for `tst` during tests; now use `config_get()` and modify `rucio.cfg` between tests

- Changed how we clear the cache in the test scripts as Travis doesn't have `netcat`/`nc`

- Improved exception handling in `sync_meta.py` to identify duplication so it's more in line with other files

- Added VO to a number of replica functions: `list_bad_replicas_status`, `get_pfn_to_rse`, `declare_bad_file_replica`, `get_bad_replicas_summary` and `get_did_from_pfns`

This is dependent on issue #60 / PR #63 as we need to store the clientside tokens by VO in order to run the successive tests.

### CODE REVIEW

**Comment by elichad** [comment]
There's a new test file `tests/test_reaper2.py` which will also need updating to the new format.

File: lib/rucio/core/meta.py [hunk 1]
```diff
@@ -63,7 +64,14 @@ def add_key(key, key_type, value_type=None, value_regexp=None, session=None):
...
     except IntegrityError as error:
-        if error.args[0] == "(IntegrityError) column key is not unique":
+        if error.args[0] == "(IntegrityError) column key is not unique" \
+           or ('UNIQUE constraint failed' in error.args[0]) \
+           or ('conflicts with persistent instance' in error.args[0]) \
+           or match('.*IntegrityError.*ORA-00001: unique constraint.*DID_KEYS_PK.*violated.*', error.args[0]) \
+           or match('.*IntegrityError.*1062.*Duplicate entry.*for key.*', error.args[0]) \
+           or match('.*IntegrityError.*duplicate key value violates unique constraint.*', error.args[0]) \
+           or match('.*UniqueViolation.*duplicate key value violates unique constraint.*', error.args[0]) \
+           or match('.*sqlite3.IntegrityError.*is not unique.*', error.args[0]):
```
Comment (elichad) [inline lib/rucio/core/meta.py:74]: Is there no way to simplify this? It is quite clumsy and could cause issues if any of these error messages change in the future. Have you asked Martin about this?

File: lib/rucio/tests/test_account.py [hunk 2]
```diff
@@ -91,8 +92,8 @@ def test_add_account_attribute(self):
 class TestAccountRestApi():
     def setup(self):
         if config_get_bool('common', 'multi_vo', raise_exception=False, default=False):
-            self.vo_header = {'X-Rucio-VO': 'tst'}
-            self.vo = {'vo': 'tst'}
+            self.vo_header = {'X-Rucio-VO': config_get('client', 'vo', raise_exception=False, default='tst')}
+            self.vo = {'vo': config_get('client', 'vo', raise_exception=False, default='tst')}
```
Comment (elichad) [inline lib/rucio/tests/test_account.py:95]: A small thing: For tidiness, I think we should avoid making multiple calls to config_get in the test setup. Instead use the VO to set the VO header. This is in quite a few files: 
- test_account.py
- test_authentication.py
- test_import_export.py
- test_multi_vo.py
- test_replica.py
- test_root_proxy.py
- test_rse.py
- test_rse_expression_parser.py
- test_subscription.py

File: lib/rucio/core/meta.py (hunk 1 above)
Comment (patrick-austin) [reply lib/rucio/core/meta.py:74]: I haven't spoken to Martin; I based this off what I saw other core functions doing (e.g. `add_dids()`). All the ones I saw use this kind of logic when excepting an `IntegrityError` and raising a `Duplicate`. Some places don't have the if clause and just raise it as a generic `RucioException`, but that won't be caught by the bootstrap process and so the second test run will fail. If there is a better way of doing it I'm happy to change it, but from looking at the code it doesn't like there is (or at least we don't do it).

File: lib/rucio/tests/test_account.py (hunk 2 above)
Comment (patrick-austin) [reply lib/rucio/tests/test_account.py:95]: Yeah that totally makes sense, I just did this with a blanket find/replace as I didn't want to change each test manually but it should definitely change as you say.

File: lib/rucio/core/meta.py (hunk 1 above)
Comment (elichad) [reply lib/rucio/core/meta.py:74]: Ah, if there is precedent elsewhere in the code, we can leave this as it is

**Review by elichad — APPROVED** [review]

**Pull request merged**